### DISCUSSION TITLE
Dump bazy psql

### DISCUSSION THREAD

**Discussion by kami33n** [AzureCommunityPL/group#412] [Pomoc]
Cześć, mam problem z dumpem bazy postgresql. Baza waży 3.5TB (zastana). Potrzebuję ja pobrać lokalnie bo tylko leży i nabija rachunek. Próbowałem podejść do sprawy standardowo i zrobić dumpa, na dysku zapisuje mi się do 1GB po czym plik nie rośnie, w metryce Network out widze, że jest duża większy ruch. Dodam, że 3.5TB to rozmiar praktycznie dwóch tabel, pozostałe ważą łącznie z 1.5GB. Ktoś może powiedzieć jak podejść do takiego dumpa i ile jaki może być koszt transferu?

**Comment by kamil-mrzyglod** [comment]
To dwa pytania:
* z jakiego regionu chcesz zaciągnąć tę bazę i czy plik chcesz zapisać u siebie czy np. na maszyncie wirtualnej w Azure albo Azure Storage
* to baza postawiona na VM czy Azure PostgreSQL

Zakładając, że robisz dumpa z DC do siebie na kompa to będzie to koszt kilkuset $ (czy 200, 300, 600 czy więcej zależy jeszcze od kilku czynników)

**Comment by kami33n** [comment]
1. Z regionu UE i plik chce zapisać u siebie lokalnie
2. Azure Postgres
3. Ok, parę stówek jest akceptowalne. Ważne, że nie kilka tys.

**Reply by kamil-mrzyglod** [reply]
No okej, a jak tego dumpa robisz? pg_dump i wskazujesz na swoją bazę w Azure? Jaki masz ustawiony compute bazy? Jakiej składni dokładnie używasz dla dumpa? Na jaki dysk zrzucasz tego dumpa (HDD / SSD)? Próbowałeś zrzucić bazę do Azure Storage zamiast na lokalny dysk?

**Reply by kami33n** [reply]
1. Używam pg_admina
2. compute - Przeznaczenie ogólne, rdzenie wirtualne: 2, 3584 GB
3. C:\Program Files\pgAdmin 4\v6\runtime\pg_dump.exe --file "Z:\\backup\\BVYLT0~J" --host "xxxx.postgres.database.azure.com" --port "5432" --username "admin@xxxx" --no-password --verbose --format=t --blobs --create --section=data "xxxxxdb"
4. dysk HDD
5. nie próbowałem zrzucać tego na Azure Storage

Zastanawia mnie czemu mimo, tak dużego ruchu wychodzącego (~40GB), na dysku zapisało się tylko około 1GB, to jest gdzieś najpierw buforowane a następnie wypychane z Azura?

**Comment by kami33n** [comment]
Podszedłem inaczej do sprawy. Stworzyłem magazyn na który robię dumpa przy użyciu Azure CLI
polecenie: pg_dump -v --xxx.database.azure.com --dbname=xxxx --user=admin  --format=d --jobs=4 --file=/usr/csuser/clouddrive/dmp/ --compress=6
tym razem zasób urósł do 8GB i dalej stoi. Podejrzewam, że urośnie jak przetworzy tabelę która waży 1.5 TB

Edit.
Po przetworzeniu 41.4 GB przestało działać, podobnie jak dzień wcześniej kiedy używałem pg_admin
Edit2:
jak mogę wykonywać dumpa w tle mimo przerwania połączenia po 20minutach z Azure CLI?

**Reply by kamil-mrzyglod** [reply]
Dump będzie się wykonywał w tle pomimo timeoutu w Cloud Shell więc tym bym się bardzo nie przejmował. Jeśli cały proces nie dostanie na twarz jakimś błędem to po jakimś czasie powinien zgrać wszystko.

Bazę masz dość małą i z mojego doświadczenia z PaaSowymi bazami są one dość kapryśne przy dużym i stałym obciążeniu (a dump bazy nie można uznać na małą i prostą operację) potrafią zrywać połączenie z X powodów. Generalnie widziałbym trzy rozwiązania aby temat popychać:
* postawienie Azure VM na którą robisz dumpa na data disk / file share - lepsza kontrola niż Cloud Shell
* postawienie Azure VM z PostgreSQL i zmigrowanie jednej bazy do drugiej a następnie pobranie bazy, do której fizycznie masz dostęp
* zrobienie tego batchami - jeśli pg_dump tego nie potrafi to jakimkolwiek SDK, które potrafi gadać z PostgreSQL i zrzucić do jakiegoś formatu archiwum w formie kilkudziesięciu / kilkuset paczek. Jeśli ma działać lokalnie to zamiast archiwizować to od razu zapisywać w lokalnej bazie

**Reply by kami33n** [reply]
postawienie Azure VM na którą robisz dumpa na data disk / file share - lepsza kontrola niż Cloud Shell - przerwało się po 130GB

postawienie Azure VM z PostgreSQL i zmigrowanie jednej bazy do drugiej a następnie pobranie bazy, do której fizycznie masz dostęp - robiąc coś takiego użył bym polecenia **pg_dump -C -h remotehost -U remoteuser dbname | psql -h localhost -U localuser dbname** więc nie wiem czy tutaj znów nie było by problemu z zerwaniem połączenia

zrobienie tego batchami - jeśli pg_dump tego nie potrafi to jakimkolwiek SDK, które potrafi gadać z PostgreSQL i zrzucić do jakiegoś formatu archiwum w formie kilkudziesięciu / kilkuset paczek. Jeśli ma działać lokalnie to zamiast archiwizować to od razu zapisywać w lokalnej bazie - do tego jeszcze nie wiem jak podejść
